### DISCUSSION TITLE
Support for programmatic layer creation (for modules)

### DISCUSSION THREAD

**Discussion by davestewart** [nuxt/nuxt#26837] [Ideas]
## Abstract

**TL;DR Nuxt config is processed and layers are finalised before modules can hook into them, meaning that if a module wants to dynamically set up a layer, potentially for another module that might rely on that layer, it's too late.**

I'm about to release version 1.4.3 of Nuxt Content Assets – which now properly supports Nuxt Image – but I had to jump through quite a few hoops to achieve it.

I wanted share my use case, discuss the hoops, and either find a better way to do what I want to do, or raise the possibility for new functionality to make the process easier / more standardised.

## Background

The old version of Nuxt Content Assets used Nuxt Image's `dir` setting to configure that Nuxt Image use the Nuxt Content Asset temp folder of assets, which was itself created at `.nuxt/content-assets` (in a similar way to how Nuxt Content creates `.nuxt/content-cache`).

Unfortunately, this prevented Nuxt Image reading from the project's default `public` folder.

So I dug into the Nuxt Image code, and noticed that [ipxSetup()](https://github.com/nuxt/image/blob/main/src/ipx.ts#L31-L37) supports additional `public` folders by way of layers.

So I figure I should be able to add layers dynamically, right?

My first idea was in effect to generate a layer dynamically by adding a `nuxt.config.json` file and `public` folder to the existing `.nuxt/content-assets` folder, but as I discovered through [quite a lot of painful debugging](https://github.com/davestewart/nuxt-content-assets/issues/76#issuecomment-2059022155) a project's layers are processed during [loadNuxt()](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/core/nuxt.ts#L523) which is way before [module setup](https://github.com/nuxt/nuxt/blob/main/packages/nuxt/src/core/nuxt.ts#L452) so this approach would not work.

And, as confirmed by @danielroe adding layers dynamically is [not yet possible](https://twitter.com/danielcroe/status/1779193275275960425).

Additionally, creating a layer folder in `.nuxt` was problematic, as in `nuxi generate` the folder itself would be rebuilt, and so every other run, the layer would either be picked up, or it wouldn't, causing the module to mysteriously break!

I then worked through a bunch of unsuccessful ideas to provide a layer folder, eventually settling on a `cache` folder in the module source itself:

```
+- nuxt_modules
    +- nuxt-content-assets
        +- cache
            +- public
            |   +- .gitkeep
            +- nuxt.config.json
```

This works, but:

- the user has to manually add this layer in their config (as module `setup()` runs too late)
- the package has to include these "dummy" folders and files (rather than there being 1st class support)
- the module playground requires additional script setup (as the module runs in `../src` rather than `node_modules`)

## Ideas for a solution

Before continuing, I just want to review the relevant Nuxt lifecycle steps as I understand it:

1. start bootstrap
2. load and process config
3. create layers
4. install modules (process hooks and run setup) in order found
5. setup imports, pages, etc
6. build nuxt

Thinking through possible approaches to find a solution...

#### Provide a hook to run setup code before nuxt processes config (such as adding a layer folder)

Nice idea, but chicken and egg with the configuration itself; nuxt reads the module from the config, so how would the module run code before it is found!?

#### Provide a pre-setup hook to allow modules to modify config or run code before the module install runs

This seems like it might be useful to set up the layer folder, rather than it being deleted in the build, but would not solve the problem of the layer being added to Nuxt, as layers are created when the config is parsed.

#### Provide a utility to allow modules to add layers dynamically such as `addLayer()`

This seems like the most obvious choice; could be as simple as passing a string to the function, same as `extends`.

It would be useful if this could be done _before_ module setup, so as described above, any modules requiring layers during setup (such as Nuxt Image) could have its layers created by other modules (such as Nuxt Content Assets).

#### Provide a way for modules to add layers through runtime folders

Perhaps also, layer folders could be added and configured simply by adding runtime folders.

I'm not sure if this goes against the programmatic nature of the Nuxt Kit API, or if a potential `addLayer()` might reference this folder.

Note also my comment about the `.nuxt` folder not always being available (this seems to differ between `dev` and `generate`) so some input about whether to use and when (in the lifecycle) to use `.nuxt` as a place of temporary storage would be welcome. 

## Proposal

I don't have an actual proposal per-se, but did want to bring up the topic for discussion anyway.

Sorry it's a bit long!

> Caveat: I would like to say my knowledge of building Nuxt Modules is patchy; I tend to learn what I need for the problem at hand, and there are probably various concepts or techniques that I could be up to speed on but I'm not.

**Comment by MatteoGabriele** [comment]
Is this proposal still being considered? I am looking for a way to programmatically add layers.
